# Lesson 2.2: Scraping Gutenberg: Filtering Queries

So far none of the work we have done on the Gutenberg catalog has not actually gotten us any texts. Instead we did the following:
- Downloaded a CSV file
- Converted it to a dataframe
- Cleaned the dataframe
- Exported as `.pickle` file

We did all this because the dataframe will used to retrieve actual texts from Gutenberg, and we want to be sure everything is clean before we modify it.

In the following lesson you will learn how to filter the dataframe for the specific values you want by running queries. Since we don't want to download all of Gutenberg, we will want to create a subset of data to download before we do the actual webscraping.

#### Load libraries

Libraries need to be reloaded for each new notebook.

In [25]:
import pandas as pd

## Import `pg_catalog_clean.pickle`


You will need to import the `.pickle` file and save it as a dataframe file.

In [26]:
df_pg_catalog = pd.read_pickle('pg_catalog_clean.pickle')

In [27]:
df_pg_catalog

text_id  type      issued  \
0            1  Text  1971-12-01   
1            2  Text  1972-12-01   
2            3  Text  1973-11-01   
3            4  Text  1973-11-01   
4            5  Text  1975-12-01   
...        ...   ...         ...   
74250    74416  Text  2024-09-14   
74251    74417  Text  2024-09-14   
74252    74418  Text  2024-09-15   
74253    74419  Text  2024-09-15   
74254    90907  Text  2019-12-12   

                                                   title language  \
0      The Declaration of Independence of the United ...       en   
1                       The United States Bill of Rights       en   
2                    John F. Kennedy's Inaugural Address       en   
3                           Lincoln's Gettysburg Address       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                subjects      locc  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                               <NA>      <NA>   
74251                                               <NA>      <NA>   
74252                                               <NA>      <NA>   
74253                                               <NA>      <NA>   
74254                                               <NA>      <NA>   

                                             bookshelves  \
0      Politics; American Revolutionary War; United S...   
1      Politics; American Revolutionary War; United S...   
2       Browsing: History - American; Browsing: Politics   
3      US Civil War; Browsing: History - American; Br...   
4      United States; Politics; American Revolutionar...   
...                                                  ...   
74250                                               <NA>   
74251                                               <NA>   
74252                                               <NA>   
74253                                               <NA>   
74254                                               <NA>   

                                           second_author      last_name  \
0                                                   <NA>      Jefferson   
1                                                   <NA>  United States   
2                                                   <NA>        Kennedy   
3                                                   <NA>        Lincoln   
4                                                   <NA>  United States   
...                                                  ...            ...   
74250                                               <NA>            Lee   
74251                                               <NA>       Brewster   
74252  Murphy, Henry Cruse, 1810-1882 [Illustrator]; ...          Verne   
74253                                               <NA>           Levy   
74254                                               <NA>           <NA>   

                      first_name birth  \
0                         Thomas  1743   
1                           <NA>  <NA>   
2      John F. (John Fitzgerald)  1917   
3                        Abraham  1809   
4                           <NA>  <NA>   
...                          .

## Filtering the dataframe

Filtering data in pandas is pretty intuitive. Essentially you select the column you want to filter and then use the method `.str.contains()` to search for the value you want to retrieve.

```python
df_pg_catalog['subjects'].str.contains('Virginia', case=False, na=False)
```
`case` indicates that the search is case sensitive. Since it doesn't matter if Virginia is capitalized or not, we can leave this as `false`.

In [28]:
df_pg_catalog['subjects'][df_pg_catalog['subjects'].str.contains('Virginia', case=False, na=False)]

5        Speeches, addresses, etc., American; United St...
72       Historical fiction; War stories; United States...
351      London (England) -- Fiction; Picaresque litera...
444      Historical fiction; War stories; United States...
2351       Virginia -- Fiction; Tobacco farmers -- Fiction
                               ...                        
69266    Villages -- Fiction; Man-woman relationships -...
69868    Orphans -- Fiction; Young women -- Fiction; Di...
70166    African Americans -- Education -- Virginia; Do...
70718    Sisters -- Fiction; Domestic fiction; Mate sel...
72725    Women dancers -- France -- Fiction; Businessme...
Name: subjects, Length: 177, dtype: string

In [29]:
df_pg_catalog[df_pg_catalog['subjects'].str.contains('Virginia', case=False, na=False) & (df_pg_catalog['language'] == 'en')]

text_id  type      issued  \
5            6  Text  1976-12-01   
72          73  Text  1993-07-01   
351        370  Text  1995-12-01   
444        463  Text  1996-03-01   
2351      2384  Text  2000-11-01   
...        ...   ...         ...   
69266    69369  Text  2022-11-17   
69868    70010  Text  2023-02-10   
70166    70331  Text  2023-03-21   
70718    70883  Text  2023-05-30   
72725    72891  Text  2024-02-07   

                                                   title language  \
5                       Give Me Liberty or Give Me Death       en   
72     The Red Badge of Courage: An Episode of the Am...       en   
351    The Fortunes and Misfortunes of the Famous Mol...       en   
444    The Red Badge of Courage: An Episode of the Am...       en   
2351   The Deliverance: A Romance of the Virginia Tob...       en   
...                                                  ...      ...   
69266                                       Deep channel       en   
69868                           The shadow between them;       en   
70166                       Educational laws of Virginia       en   
70718                         Doctor Hathern's daughters       en   
72725                                      The wild fawn       en   

                                                subjects  locc  \
5      Speeches, addresses, etc., American; United St...  E201   
72     Historical fiction; War stories; United States...    PS   
351    London (England) -- Fiction; Picaresque litera...    PR   
444    Historical fiction; War stories; United States...    PS   
2351     Virginia -- Fiction; Tobacco farmers -- Fiction    PS   
...                                                  ...   ...   
69266  Villages -- Fiction; Man-woman relationships -...    PS   
69868  Orphans -- Fiction; Young women -- Fiction; Di...    PS   
70166  African Americans -- Education -- Virginia; Do...    LC   
70718  Sisters -- Fiction; Domestic fiction; Mate sel...    PS   
72725  Women dancers -- France -- Fiction; Businessme...    PS   

                                             bookshelves second_author  \
5      American Revolutionary War; Browsing: History ...          <NA>   
72     US Civil War; Historical Fiction; Best Books E...          <NA>   
351    Browsing: Culture/Civilization/Society; Browsi...          <NA>   
444    US Civil War; Historical Fiction; Bestsellers,...          <NA>   
2351   Bestsellers, American, 1895-1923; Browsing: Cu...          <NA>   
...                                                  ...           ...   
69266  Browsing: Culture/Civilization/Society; Browsi...          <NA>   
69868  Browsing: Culture/Civilization/Society; Browsi...          <NA>   
70166  Browsing: Culture/Civilization/Society; Browsi...          <NA>   
70718  Browsing: Culture/Civilization/Society; Browsi...          <NA>   
72725  Browsing: Culture/Civilization/Society; Browsi...          <NA>   

      last_name              first_name       birth death  
5         Henry                 Patrick        1736  1799  
72        Crane                 Stephen        1871  1900  
351       Defoe                  Daniel       1661?  1731  
444       Crane                 Stephen        1871  1900  
2351    Glasgow  Ellen Anderson Gholson        1873  1945  
...         ...                     ...         ...   ...  
69266  Montague       Margaret Prescott        1878  1955  
69868    Miller           Alex. McVeigh  Mrs., 1850  1937  
70166  Douglass     Margaret Crittenden        1822        
70718    Holmes               Mary Jane        1825  1907  
72725    Taylor              Mary Imlay        1878  1938  

[176 rows x 13 columns]

This resulted in 117 records. We may remember that Gutenberg is about 70k records so we have reduced this down quite a bit.

We can try to narrow this down a bit by excluding all works that are not in English `en`

```python
df_pg_catalog[df_pg_catalog['subjects'].str.contains('Virginia', case=False, na=False) & (df_pg_catalog['language'] == 'en')]
```

Instead of seeing if the string contains `en`, now we are simply asking if it equals `en`.

In [30]:
df_pg_catalog[df_pg_catalog['subjects'].str.contains('Virginia', case=False, na=False) & (df_pg_catalog['language'] == 'en')]

text_id  type      issued  \
5            6  Text  1976-12-01   
72          73  Text  1993-07-01   
351        370  Text  1995-12-01   
444        463  Text  1996-03-01   
2351      2384  Text  2000-11-01   
...        ...   ...         ...   
69266    69369  Text  2022-11-17   
69868    70010  Text  2023-02-10   
70166    70331  Text  2023-03-21   
70718    70883  Text  2023-05-30   
72725    72891  Text  2024-02-07   

                                                   title language  \
5                       Give Me Liberty or Give Me Death       en   
72     The Red Badge of Courage: An Episode of the Am...       en   
351    The Fortunes and Misfortunes of the Famous Mol...       en   
444    The Red Badge of Courage: An Episode of the Am...       en   
2351   The Deliverance: A Romance of the Virginia Tob...       en   
...                                                  ...      ...   
69266                                       Deep channel       en   
69868                           The shadow between them;       en   
70166                       Educational laws of Virginia       en   
70718                         Doctor Hathern's daughters       en   
72725                                      The wild fawn       en   

                                                subjects  locc  \
5      Speeches, addresses, etc., American; United St...  E201   
72     Historical fiction; War stories; United States...    PS   
351    London (England) -- Fiction; Picaresque litera...    PR   
444    Historical fiction; War stories; United States...    PS   
2351     Virginia -- Fiction; Tobacco farmers -- Fiction    PS   
...                                                  ...   ...   
69266  Villages -- Fiction; Man-woman relationships -...    PS   
69868  Orphans -- Fiction; Young women -- Fiction; Di...    PS   
70166  African Americans -- Education -- Virginia; Do...    LC   
70718  Sisters -- Fiction; Domestic fiction; Mate sel...    PS   
72725  Women dancers -- France -- Fiction; Businessme...    PS   

                                             bookshelves second_author  \
5      American Revolutionary War; Browsing: History ...          <NA>   
72     US Civil War; Historical Fiction; Best Books E...          <NA>   
351    Browsing: Culture/Civilization/Society; Browsi...          <NA>   
444    US Civil War; Historical Fiction; Bestsellers,...          <NA>   
2351   Bestsellers, American, 1895-1923; Browsing: Cu...          <NA>   
...                                                  ...           ...   
69266  Browsing: Culture/Civilization/Society; Browsi...          <NA>   
69868  Browsing: Culture/Civilization/Society; Browsi...          <NA>   
70166  Browsing: Culture/Civilization/Society; Browsi...          <NA>   
70718  Browsing: Culture/Civilization/Society; Browsi...          <NA>   
72725  Browsing: Culture/Civilization/Society; Browsi...          <NA>   

      last_name              first_name       birth death  
5         Henry                 Patrick        1736  1799  
72        Crane                 Stephen        1871  1900  
351       Defoe                  Daniel       1661?  1731  
444       Crane                 Stephen        1871  1900  
2351    Glasgow  Ellen Anderson Gholson        1873  1945  
...         ...                     ...         ...   ...  
69266  Montague       Margaret Prescott        1878  1955  
69868    Miller           Alex. McVeigh  Mrs., 1850  1937  
70166  Douglass     Margaret Crittenden        1822        
70718    Holmes               Mary Jane        1825  1907  
72725    Taylor              Mary Imlay        1878  1938  

[176 rows x 13 columns]

This gives us 176 values. This is still too many.

One of the issus Moretti ran into is the ambiguity of fiction. We can perhaps avoid that ambiguity by excluding those types of books from our corpus. We can do this by introducing yet another query line that looks for `fiction` but then adding a `~` in front of it to negate the result, i.e. **do not** include anything that has the word ficiton in it.


In [31]:
df_pg_catalog[
    df_pg_catalog['subjects'].str.contains('Virginia', case=False, na=False) & 
    (df_pg_catalog['language'] == 'en') & 
    ~df_pg_catalog['subjects'].str.contains('fiction', case=False, na=False)
]

text_id  type      issued  \
5            6  Text  1976-12-01   
2637      2674  Text  2001-06-01   
2858      2898  Text  2001-11-01   
3085      3126  Text  2004-10-10   
4206      4247  Text  2003-07-01   
...        ...   ...         ...   
64913    64992  Text  2021-04-05   
64948    65027  Text  2021-04-08   
65081    65160  Text  2021-04-25   
67666    67745  Text  2022-03-31   
70166    70331  Text  2023-03-21   

                                                   title language  \
5                       Give Me Liberty or Give Me Death       en   
2637   The Complete Writings of Charles Dudley Warner...       en   
2858   Pioneers of the Old South: A Chronicle of Engl...       en   
3085                                        On Horseback       en   
4206   A Briefe and True Report of the New Found Land...       en   
...                                                  ...      ...   
64913                       Narrative of Henry Box Brown       en   
64948  A narrative of some remarkable incidents, in t...       en   
65081                    The Discoveries of John Lederer       en   
67666                 Yorktown: Climax of the Revolution       en   
70166                       Educational laws of Virginia       en   

                                                subjects        locc  \
5      Speeches, addresses, etc., American; United St...        E201   
2637   Autobiographies; Virginia -- Description and t...          PS   
2858   Southern States -- History -- Colonial period,...  E151; F206   
3085   California -- Description and travel; Virginia...        F206   
4206   Indians of North America -- North Carolina; Ro...        F206   
...                                                  ...         ...   
64913  African American abolitionists -- Biography; B...        E300   
64948  Fugitive slaves -- United States -- Biography;...    E300; HT   
65081  Indians of North America -- North Carolina -- ...        F206   
67666  Virginia -- History -- Revolution, 1775-1783; ...        E201   
70166  African Americans -- Education -- Virginia; Do...          LC   

                                             bookshelves  \
5      American Revolutionary War; Browsing: History ...   
2637   Browsing: Biographies; Browsing: Literature; B...   
2858   United States; Children's History; Browsing: H...   
3085   Browsing: History - American; Browsing: Travel...   
4206   Browsing: History - American; Browsing: Histor...   
...                                                  ...   
64913  Browsing: Biographies; Browsing: Culture/Civil...   
64948  Browsing: Biographies; Browsing: Culture/Civil...   
65081  Browsing: History - American; Browsing: Travel...   
67666  Browsing: History - American; Browsing: Histor...   
70166  Browsing: Culture/Civilization/Society; Browsi...   

                                        second_author last_name  \
5                                                <NA>     Henry   
2637                                             <NA>    Warner   
2858               Johnson, Allen, 1870-1931 [Editor]  Johnston   
3085                                             <NA>    Warner   
4206                                             <NA>   Harriot   
...                                               ...       ...   
64913   Stearns, Charles (Abolitionist) [Contributor]     Brown   
64948  Hurnard, Robert [Author of introduction, etc.]    Bayley   
65081        Talbot, William, Sir, -1691 [Translator]   Lederer   
67666           Pitkin, Thomas M., 1901-1988 [Editor]     Hatch   
70166                                            <NA>  Douglass   

                first_name                                  birth  \
5                  Patrick                                   1736   
2637        Charles Dudley                                   1829   
2858                  Mary                                   1870   
3085        Charles Dudley                                   1829   
4206     

This is a bit better since we are now down to 99 texts. We can get even more finegrained by excluding multiple words. There are also speeches in the data set. We can exclude these without having to write a whole new line of code. Instead we can use the `|` or operator to exclude `fiction` or `speeches`. Since we want to exclude both `speech` and `speeches` we can use a wildcard `*` on `speech*` to achieve this effect.

In [32]:
df_pg_catalog[
    df_pg_catalog['subjects'].str.contains('Virginia', case=False, na=False) & 
    (df_pg_catalog['language'] == 'en') & 
    ~df_pg_catalog['subjects'].str.contains('fiction|speech*', case=False, na=False)
]

text_id  type      issued  \
2637      2674  Text  2001-06-01   
2858      2898  Text  2001-11-01   
3085      3126  Text  2004-10-10   
4206      4247  Text  2003-07-01   
4721      4762  Text  2003-12-01   
...        ...   ...         ...   
64913    64992  Text  2021-04-05   
64948    65027  Text  2021-04-08   
65081    65160  Text  2021-04-25   
67666    67745  Text  2022-03-31   
70166    70331  Text  2023-03-21   

                                                   title language  \
2637   The Complete Writings of Charles Dudley Warner...       en   
2858   Pioneers of the Old South: A Chronicle of Engl...       en   
3085                                        On Horseback       en   
4206   A Briefe and True Report of the New Found Land...       en   
4721                        Civil Government of Virginia       en   
...                                                  ...      ...   
64913                       Narrative of Henry Box Brown       en   
64948  A narrative of some remarkable incidents, in t...       en   
65081                    The Discoveries of John Lederer       en   
67666                 Yorktown: Climax of the Revolution       en   
70166                       Educational laws of Virginia       en   

                                                subjects        locc  \
2637   Autobiographies; Virginia -- Description and t...          PS   
2858   Southern States -- History -- Colonial period,...  E151; F206   
3085   California -- Description and travel; Virginia...        F206   
4206   Indians of North America -- North Carolina; Ro...        F206   
4721                 Virginia -- Politics and government          JK   
...                                                  ...         ...   
64913  African American abolitionists -- Biography; B...        E300   
64948  Fugitive slaves -- United States -- Biography;...    E300; HT   
65081  Indians of North America -- North Carolina -- ...        F206   
67666  Virginia -- History -- Revolution, 1775-1783; ...        E201   
70166  African Americans -- Education -- Virginia; Do...          LC   

                                             bookshelves  \
2637   Browsing: Biographies; Browsing: Literature; B...   
2858   United States; Children's History; Browsing: H...   
3085   Browsing: History - American; Browsing: Travel...   
4206   Browsing: History - American; Browsing: Histor...   
4721   Browsing: History - American; Browsing: Politi...   
...                                                  ...   
64913  Browsing: Biographies; Browsing: Culture/Civil...   
64948  Browsing: Biographies; Browsing: Culture/Civil...   
65081  Browsing: History - American; Browsing: Travel...   
67666  Browsing: History - American; Browsing: Histor...   
70166  Browsing: Culture/Civilization/Society; Browsi...   

                                        second_author last_name  \
2637                                             <NA>    Warner   
2858               Johnson, Allen, 1870-1931 [Editor]  Johnston   
3085                                             <NA>    Warner   
4206                                             <NA>   Harriot   
4721                                             <NA>       Fox   
...                                               ...       ...   
64913   Stearns, Charles (Abolitionist) [Contributor]     Brown   
64948  Hurnard, Robert [Author of introduction, etc.]    Bayley   
65081        Talbot, William, Sir, -1691 [Translator]   Lederer   
67666           Pitkin, Thomas M., 1901-1988 [Editor]     Hatch   
70166                                            <NA>  Douglass   

                first_name                                  birth  \
2637        Charles Dudley                                   1829   
2858                  Mary                                   1870   
3085        Charles Dudley                                   1829   
4206                Thomas                                   1560   
4721     

We can save all this as a new dataframe.

In [33]:
df_virginia_history = df_pg_catalog[
    df_pg_catalog['subjects'].str.contains('Virginia', case=False, na=False) & 
    (df_pg_catalog['language'] == 'en') & 
    ~df_pg_catalog['subjects'].str.contains('fiction|speech*', case=False, na=False)
]

We can use our .pickle() function to export this.

In [34]:
df_virginia_history.to_pickle('virginia_history.pickle')